servicetalk-traffic-resilience-http/src/main/java/io/servicetalk/traffic/resilience/http/AbstractTrafficResilienceHttpFilter.java

```diff
@@ -194,11 +199,29 @@ Ticket wrapTicket(@Nullable final ServerListenContext serverListenContext, final
         return ticket;
     }
 
+    private Single<StreamingHttpResponse> doHandleLocalCapacityRejection(
+            Function<StreamingHttpRequest, Single<StreamingHttpResponse>> delegate,
+            @Nullable ServerListenContext serverListenContext,
+            StreamingHttpRequest request,
+            @Nullable StreamingHttpResponseFactory responseFactory) {
+        return dryRun ? handlePassthrough(delegate, request) :
+                handleLocalCapacityRejection(serverListenContext, request, responseFactory);
```

Comment:
In my understanding, the "dry-run" is a mode when we are collecting information about decisions made (like rejections), but don't actually enforce those decisions. In the current impl, we always call `handlePassthrough` without collecting any data of local rejections that could happen if we disable "dry-run" mode (the fact that we want to reject with the cause why and the limiter state).

Should we log or track any counters for rejection causes? If we log, how do we make sure we won't overflow the logs consumer in case all requests are getting rejections? Can we log once in 24 hours with summary stats?

Reply:
Let's consider specific cases to make the discussion concrete.
First there is `doHandleLocalCapacityRejection` which is used twice. In both of those cases the ticket either didn't exist or we had another breaker that we tested that wants to short circuit. In both cases the observer will see the same thing dry run or not, so we should have data on the state. The only difference is that in dry-run mode we pass-through to do the request as normal after that whereas in non-dry run we defer to concrete implementations to create an exception to return back to the user.
The breaker rejection case is the same other than there is only one use of it.

We can certainly add some logging if we think the observers are not enough that includes the state that is relevant to the rejection cause. I'll plan to add that, but please let me know if I've miss the mark here.

Reply:
Doh, I missed the `TrafficResiliencyObserver` part. You are right, as long as we have a good observer impl, we should be able to capture all details about rejection cause. The only thing that is hard to get from the observer is the current internal state of the `CapacityLimiter` (partition) at the time of `onRejectedLimit`.

How do you think, should we provide something like "logging observer" or "debugging observer" OOTB?

Reply:
It looks like the various flavors of CapacityLimiter each have their own observers. Maybe they're the right place to do those observations.